### PR TITLE
Fix SubtypingDiscoverer with shared i31

### LINKED ISSUE #7773: Infinite loop in Unsubtyping
Reduced to
```wat
(module
 (rec
  (type $0 (shared (func (result (ref null $22)))))
  (type $1 (struct (field i8) (field (ref null $9)) (field (mut i31ref))))
  (type $2 (sub (shared (array (mut (ref $26))))))
  (type $3 (sub (array (mut (ref null $17)))))
  (type $4 (sub (struct (field i16) (field (ref null $15)) (field f32) (field (mut (ref null $0))))))
  (type $5 (sub (struct (field i16))))
  (type $6 (func (param f32 (ref $12) i32 v128) (result f32)))
  (type $7 (sub (array v128)))
  (type $8 (sub (shared (array (mut i16)))))
  (type $9 (sub (array (mut (ref null (shared extern))))))
  (type $10 (shared (struct (field (mut (ref $18))))))
  (type $11 (sub (func (param (ref $14) (ref $13) (ref struct) v128 f64 (ref null (shared i31)) i32) (result f64))))
  (type $12 (struct (field (mut i64)) (field (mut f32))))
  (type $13 (sub (func (result (ref $33)))))
  (type $14 (sub (struct (field (mut i8)) (field (mut f64)) (field i32))))
  (type $15 (sub (shared (array (mut i32)))))
  (type $16 (array i32))
  (type $17 (sub (shared (struct (field f64) (field (mut (ref $27))) (field (ref $25))))))
  (type $18 (shared (array i8)))
  (type $19 (func (param i32 (ref $6) (ref null $5) (ref null $18) (ref null $10) (ref null $20) (ref (shared i31)) (ref null (shared any))) (result (ref null $6))))
  (type $20 (shared (func (param (ref null $21)) (result (ref $7)))))
  (type $21 (sub (shared (func (result (ref $18))))))
  (type $22 (sub (shared (struct (field (mut (ref $32)))))))
  (type $23 (sub (struct (field structref) (field (mut f32)) (field (mut (ref null (shared func)))) (field f64) (field (mut f64)) (field (ref null $16)))))
  (type $24 (shared (struct)))
  (type $25 (sub (shared (func (param i64 v128 (ref null $18)) (result f64)))))
  (type $26 (shared (array i16)))
  (type $27 (sub $8 (shared (array (mut i16)))))
  (type $28 (sub $17 (shared (struct (field f64) (field (mut (ref $27))) (field (ref $29))))))
  (type $29 (sub final $25 (shared (func (param i64 v128 (ref null $18)) (result f64)))))
  (type $30 (sub final $11 (func (param (ref struct) (ref null $13) structref v128 f64 (ref null (shared eq)) i32) (result f64))))
  (type $31 (sub final $27 (shared (array (mut i16)))))
  (type $32 (shared (array (mut i16))))
  (type $33 (sub $11 (func (param eqref (ref null $13) structref v128 f64 (ref null (shared eq)) i32) (result f64))))
  (type $34 (sub (struct)))
  (type $35 (sub (shared (struct))))
  (type $36 (sub $14 (struct (field (mut i8)) (field (mut f64)) (field i32))))
  (type $37 (sub final $23 (struct (field nullref) (field (mut f32)) (field (mut (ref null (shared func)))) (field f64) (field (mut f64)) (field (ref $16)))))
  (type $38 (sub (shared (func (param (ref null $1) (ref null $0)) (result (ref null $12))))))
 )
 (type $39 (func (param i64 (ref string) i64) (result (ref $11))))
 (type $40 (func (result (ref null $24))))
 (type $41 (func (param (ref null (shared i31)))))
 (elem declare func $3)
 (export "func_273" (func $1))
 (func $0 (type $39) (param $0 i64) (param $1 (ref string)) (param $2 i64) (result (ref $11))
  (ref.func $3)
 )
 (func $1 (type $40) (result (ref null $24))
  (unreachable)
 )
 (func $2 (type $41) (param $0 (ref null (shared i31)))
  (drop
   (i31.get_s
    (local.get $0)
   )
  )
 )
 (func $3 (type $33) (param $0 eqref) (param $1 (ref null $13)) (param $2 structref) (param $3 v128) (param $4 f64) (param $5 (ref null (shared eq))) (param $6 i32) (result f64)
  (unreachable)
 )
)
```
```
bin/wasm-opt a.wat -all --unsubtyping
```
seems to hang forever. This is not 100% deterministic, but it actually became more so as I reduced, and seems quite reliable on this testcase (and on two different Linux machines).

### PR DESCRIPTION
**Pull request by tlively**
SubtypingDiscoverer previously noted that the child of I31Get
expressions were subtypes of i31, but they can alternatively be subtypes
of (shared i31) and this was not handled. This should properly be
modeled as a new kind of type constraint in SubtypingDiscoverer, but the
more robust solution would be to refactor SubtypingDiscoverer to use
principal types so the sharedness can be a type variable. For now, just
look at the child to figure out what shareness to use.

This bug was found because it caused an infinite loop in subtyping, but
add assertions that would have caught it directly and add a simpler test
that would have triggered the assertions had they existed before.

As drive-bys, add a new `insert` overload to InsertOrderedSet to make
the compiler happy with UNSUBTYPING_DEBUG set and make the whitespace in
another unsubtyping test consistent with the whitespace in the rest of
the file.

Fixes #7773.

### CODE REVIEW

File: src/ir/subtype-exprs.h [hunk 1]
```diff
@@ -273,7 +273,15 @@ struct SubtypingDiscoverer : public OverriddenVisitor<SubType> {
   void visitI31Get(I31Get* curr) {
     // This could be |noteNonFlowSubtype| but as there are no subtypes of i31
     // it does not matter.
-    self()->noteSubtype(curr->i31, Type(HeapType::i31, Nullable));
+    // TODO: This should by i31 with a sharedness type variable, but we will
```
Comment (kripken) [inline src/ir/subtype-exprs.h]:
```suggestion
    // TODO: This should be i31 with a sharedness type variable, but we will
```

**Review by kripken — APPROVED** [review]
